solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -1,26 +1,81 @@
+use anchor_lang::prelude::{CpiContext, Pubkey};
 use anchor_lang::solana_program::msg;
+use anchor_spl::token::{spl_token, Burn, MintTo, Transfer};
 use ibc::applications::transfer::context::{
     TokenTransferExecutionContext, TokenTransferValidationContext,
 };
 use ibc::applications::transfer::error::TokenTransferError;
 use ibc::applications::transfer::PrefixedCoin;
 use ibc::core::ics24_host::identifier::{ChannelId, PortId};
 use ibc::Signer;
+use uint::FromDecStrErr;
 
 // use crate::module_holder::IbcStorage<'_,'_>;
-use crate::IbcStorage;
+use crate::{id, IbcStorage, MINT_ESCROW_SEED};
 
-impl TokenTransferExecutionContext for IbcStorage<'_, '_> {
+impl TokenTransferExecutionContext for IbcStorage<'_, '_, '_, '_> {
     fn send_coins_execute(
         &mut self,
-        _from: &Self::AccountId,
-        _to: &Self::AccountId,
-        _amt: &PrefixedCoin,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &PrefixedCoin,
     ) -> Result<(), TokenTransferError> {
-        //let sender_id = from.to_string();
-        //let receiver_id = to.to_string();
-        //let base_denom = amt.denom.base_denom.to_string();
-        todo!()
+        msg!(
+            "Sending coins from account {} to account {}, trace path {}, base \
+             denom {}",
+            from,
+            to,
+            amt.denom.trace_path,
+            amt.denom.base_denom
+        );
+        let sender_id = from.to_string();
+        let receiver_id = to.to_string();
```

Comment:
I’m quite suspicious of those and other `to_string` calls in this function. For example comparison in `find` later should just work with no string conversion.

Reply:
We have to convert `AccountId` from IBC to string since its just a [wrapper over the string.](https://github.com/cosmos/ibc-rs/blob/main/crates/ibc/src/applications/transfer/context.rs#L25). But for comparing the Pubkeys, we can just do it directly instead of converting.

Reply:
or better yet, we convert the `AccountId` to string and then to Pubkey. So all comparisions happen with Pubkey and the conversion happens only once.
```suggestion
        let sender_id = Pubkey::from_str(&from.to_string()).unwrap();
        let receiver_id = Pubkey::from_str(&to.to_string()).unwrap();
```

Reply:
So that would mean we don’t need to do conversion when we iterate, right? Could we do `as_str()` instead of `to_string()`?

Reply:
Yes we dont do conversion when we iterate. I dont think we can use `as_str` as the `ibc::Signer` doesnt implement it.

Reply:
It has `as_ref` though.

It might be worth looking into using `Pubkey` as `Self::AccountId`.  Or some kind of wrapper around `Pubkey` since we need a type which has `TryFrom<ibc::Signer>`.

Reply:
sure sure